rfc/20241021-dynamic-provider-instances.md

```diff
@@ -0,0 +1,766 @@
+
+Specifically:
+
+- [`AbsProviderConfig`](https://pkg.go.dev/github.com/opentofu/opentofu@v1.8.3/internal/addrs#AbsProviderConfig) would be renamed to `ConfigProviderConfig` and would now represent a not-yet-expanded provider configuration.
+
+    (The two appearances of the word "Config" in the name are unfortunate but consistent with the existing naming scheme elsewhere in `package addrs`, where the `Config` prefix represents the address of a static configuration block while the `Abs` prefix with the same suffix represents the potentially-many instances it "expands" into. `Config`-prefixed addresses typically have a `Module Module` field, whiile `Abs`-prefixed addresses typically have a `Module ModuleInstance` field.)
+- A new type `AbsProviderInstance` represents a fully-expanded provider instance address:
+
+    ```go
+    type AbsProviderInstance struct {
+        Module   ModuleInstance
+        Provider Provider
+        Alias    string
+        Key      InstanceKey
+    }
+    ```
+
+    This type is the most specific type and would become the lookup key in the language runtime's internal map of the `providers.Interface` objects for each active provider instance.
+- [`LocalProviderConfig`](https://pkg.go.dev/github.com/opentofu/opentofu@v1.8.3/internal/addrs#LocalProviderConfig) remains but its scope is reduced only to describing which provider configuration block each `resource` or `data` block is associated with during graph construction, so that we can add the appropriate dependency edges.
+- A new type `LocalProviderInstance` represents the local-scoped equivalent of `AbsProviderInstance` as described above:
+
+    ```go
+    type LocalProviderInstance struct {
+        LocalName string
+        Alias     string
+        Key       InstanceKey
+    }
+    ```
+
+    An address of this type is the result of fully-evaluating the expression in the `provider` argument of a `resource` or `data` block, once the instance key expression has been evaluated. An address of this type can then in turn be translated into an `AbsProviderInstance` using the local provider name tables in each module to finally look up the correct `providers.Interface` to use for actions related to a resource instance.
+- The interface type [`ProviderConfig`](https://pkg.go.dev/github.com/opentofu/opentofu@v1.8.3/internal/addrs#ProviderConfig) is used only to deal with the annoying situation during graph construction where configuration-based graph nodes produce "local" addresses while state-based and plan-based graph nodes produce "absolute" addresses.
```

Comment:
based on my prototyping, the ProviderConfig type is no longer truly necessary in the codebase with the ProviderTransformer refactoring.